src/array/cuda/labor_sampling.cu

```diff
@@ -0,0 +1,807 @@
+  auto ds_arr = weights ? NewFloatArray(num_rows, ctx, sizeof(FloatType) * 8) : NullArray();
+  auto ds = static_cast<FloatType*>(ds_arr->data);
+  // d2s stands for (A^2)_{*s} in arXiv:2210.13339, ^2 is elementwise.
+  auto d2s_arr = weights ? NewFloatArray(num_rows, ctx, sizeof(FloatType) * 8) : NullArray();
+  auto d2s = static_cast<FloatType*>(d2s_arr->data);
+
+  if (weights) {
+    auto b_offsets = thrust::make_transform_iterator(rows, IndptrFunc<IdType>{indptr});
+    auto e_offsets = thrust::make_transform_iterator(rows, IndptrFunc<IdType>{indptr + 1});
+
+    auto A_A2 = thrust::make_transform_iterator(A, SquareFunc<FloatType>{});
+    auto ds_d2s = thrust::make_zip_iterator(ds, d2s);
+
+    size_t prefix_temp_size = 0;
+    CUDA_CALL(cub::DeviceSegmentedReduce::Reduce(
+      nullptr, prefix_temp_size,
+      A_A2,
+      ds_d2s,
+      num_rows,
+      b_offsets,
+      e_offsets,
+      TupleSum{},
+      thrust::make_tuple((FloatType)0, (FloatType)0),
+      stream));
+    auto temp = allocator.alloc_unique<char>(prefix_temp_size);
+    CUDA_CALL(cub::DeviceSegmentedReduce::Reduce(
+      temp.get(), prefix_temp_size,
+      A_A2,
+      ds_d2s,
+      num_rows,
+      b_offsets,
+      e_offsets,
+      TupleSum{},
+      thrust::make_tuple((FloatType)0, (FloatType)0),
+      stream));
+  }
+
+  thrust::counting_iterator<IdType> iota(0);
+  thrust::for_each(exec_policy, iota, iota + num_rows, DegreeFunc<IdType, FloatType>{
+    (IdType)num_picks, rows, indptr, weights ? ds : nullptr, in_deg.get(), cs});
+
+  // fill subindptr
+  auto subindptr_arr = NewIdArray(num_rows + 1, ctx, sizeof(IdType) * 8);
+  auto subindptr = static_cast<IdType*>(subindptr_arr->data);
+
+  IdType hop_size;
+  {
+    size_t prefix_temp_size = 0;
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(nullptr, prefix_temp_size,
+      in_deg.get(),
+      subindptr,
+      num_rows + 1,
+      stream));
+    auto temp = allocator.alloc_unique<char>(prefix_temp_size);
+    CUDA_CALL(cub::DeviceScan::ExclusiveSum(temp.get(), prefix_temp_size,
+      in_deg.get(),
+      subindptr,
+      num_rows + 1,
+      stream));
+
+    device->CopyDataFromTo(subindptr, num_rows * sizeof(hop_size), &hop_size, 0,
+      sizeof(hop_size),
+      ctx,
+      DGLContext{kDGLCPU, 0},
+      mat.indptr->dtype);
+  }
+  auto hop_arr = NewIdArray(hop_size, ctx, sizeof(IdType) * 8);
+  CSRMatrix smat(num_rows, mat.num_cols, subindptr_arr, hop_arr, NullArray(), mat.sorted);
+  // Consider fusing CSRToCOO into StencilOpFused kernel
+  auto smatcoo = CSRToCOO(smat, false);
+
+  auto idx_coo_arr = smatcoo.row;
+  auto idx_coo = static_cast<IdType*>(idx_coo_arr->data);
+
+  auto hop_1 = static_cast<IdType*>(hop_arr->data);
+  auto rands = allocator.alloc_unique<FloatType>(importance_sampling ? hop_size : 1);
+  auto probs_found = allocator.alloc_unique<FloatType>(importance_sampling ? hop_size : 1);
+
+  if (weights) {  // the computed c values are wrong when there are weights, so we recompute.
```

Comment:
This code organization (especially with the description saying "computed c values are wrong") is confusing.  Could we instead do
```C++
// compute c_s
if (weights) {
  // this block...
  const expr int BLOCK_CTAS = BLOCK_SIZE / CTA_SIZE;
  // ...
} else {
  // Move L596-L597 here
  // ...
}
```

Reply:
I don't want to launch a kernel just to compute c_s when there are no weights. It would make the performance worse.

Reply:
I opted for performance mostly, not code clarity. That is why some things may look out of place but I gave a lot of thought into how I can make things fast, expecially the importance_sampling == 0 case.

Reply:
My goal was to launch the fewest number of kernels possible.

Reply:
The kernel launch that inits cs does other things too, so I can't move it here.

Reply:
I see.  That makes sense.  To make the code clearer you can move the if block below L596-L597 so the code that does one thing (in this case computing `c_s`) are grouped together:
```C++
// If the graph is unweighted, we can compute c_s and in_deg in one kernel as c_s = k / in_deg.
// However, if the graph is weighted, we have to compute in_deg and c_s in separate kernels.
// L596-L597...
// L636-L655...
```